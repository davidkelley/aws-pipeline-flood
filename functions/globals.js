/**
* Determines the level of logging that the function will output. This enables
* us to reduce the amount of "noise" generated by a battle-tested function,
* whilst enabling us to output additional logging on a new version.
*
* Setting `process.env.LOG_LEVEL` will overwrite the default value of `"info"`
* for this variable.
*
* @type {String}
*/
export const LOG_LEVEL = (process.env.LOG_LEVEL || 'info');

/**
 * The AWS Region that this Lambda function is currently executing in.
 *
 * _Note: This is normally populated by the Lambda execution environment._
 *
 * @type {String}
 */
export const AWS_REGION = process.env.AWS_REGION;


/**
 * The secret API Key used to communicate with the Flood.IO account
 *
 * @type {String}
 */
export const FLOOD_API_KEY = process.env.FLOOD_API_KEY;

/**
 * The URL used to communicate with the Flood.IO API
 *
 * @type {String}
 */
export const FLOOD_API_URI = 'https://api.flood.io/';

/**
 * The path used for sending requests to action Floods
 *
 * @type {String}
 */
export const FLOODS_PATH = 'floods';

/**
 * The path used for sending requests to action Grids
 *
 * @type {String}
 */
export const GRIDS_PATH = 'grids';
