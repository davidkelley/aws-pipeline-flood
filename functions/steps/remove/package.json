{
  "name": "remove",
  "version": "0.0.0",
  "private": true,
  "main": "./remove.js"
}
