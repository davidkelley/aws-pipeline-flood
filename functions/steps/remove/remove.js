import request from 'request-promise-native';
import { FLOOD_API_KEY, FLOOD_API_URI, GRIDS_PATH } from '../../globals';

/**
 * Removes a flood grid
 *
 * See: https://github.com/flood-io/api-docs/blob/master/endpoints/grids/DELETE-grids-grid_id.md
 *
 * Callback function will be called with an object containing the new status of the grid
 *
 * @param {Object} event - object containing the grid uuid
 * @param {Object} context - Lambda runtime information
 * @param {Function} callback - function that will be invoked with (err) or (null, response)
 */
export function handler(
  event = {},
  context,
  callback = () => {}
) {
  const { grid: { uuid } } = event;
  const options = {
    uri: `${FLOOD_API_URI}${GRIDS_PATH}/${uuid}`,
    auth: {
      user: FLOOD_API_KEY,
    },
  };

  request.delete(options)
    .then(res => {
      callback(null, Object.assign({}, event, JSON.parse(res)));
    })
    .catch(err => {
      callback(err);
    });
}
