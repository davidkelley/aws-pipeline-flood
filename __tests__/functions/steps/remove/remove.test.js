/* eslint-disable import/no-unresolved */

import request from 'request-promise-native';
import { handler } from '@functions/steps/remove/remove';
import nock from 'nock';
import { FLOOD_API_URI, GRIDS_PATH } from '@functions/globals';

const mockResponse = {
  result: 'mock',
};
const mockUUID = '123123';

let requestSpy;
let callbackSpy;

beforeAll(() => {
  nock.disableNetConnect();
  requestSpy = jest.spyOn(request, 'delete');
});

afterAll(() => {
  nock.enableNetConnect();
});

describe('Remove grid suite', () => {
  describe('Handles successful query', () => {
    beforeEach((done) => {
      nock(`${FLOOD_API_URI}${GRIDS_PATH}`)
        .delete(`/${mockUUID}`)
        .reply(200, mockResponse, { 'Content-Type': 'application/json' });

      callbackSpy = jest.fn().mockImplementation(() => {
        done();
      });

      handler({
        grid: {
          uuid: mockUUID,
        },
      }, null, callbackSpy);
    });

    describe('Calling handler', () => {
      it('Called endpoint to remove grid', () => {
        expect(requestSpy.mock.calls).toMatchSnapshot();
      });

      it('Invoked callback', () => {
        expect(callbackSpy.mock.calls).toMatchSnapshot();
      });
    });
  });

  describe('Handles unsuccesful response ', () => {
    beforeEach((done) => {
      nock(`${FLOOD_API_URI}${GRIDS_PATH}`)
        .delete(`/${mockUUID}`)
        .reply(404);

      callbackSpy = jest.fn().mockImplementation(() => {
        done();
      });

      handler({
        grid: {
          uuid: mockUUID,
        },
      }, null, callbackSpy);
    });

    describe('Calling handler', () => {
      it('Returned the error to the callback', () => {
        expect(callbackSpy.mock.calls).toMatchSnapshot();
      });
    });
  });
});
