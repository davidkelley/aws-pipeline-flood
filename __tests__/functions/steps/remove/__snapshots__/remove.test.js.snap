// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Remove grid suite Handles successful query Calling handler Called endpoint to remove grid 1`] = `
Array [
  Array [
    Object {
      "auth": Object {
        "user": "mock-api-key",
      },
      "uri": "https://api.flood.io/grids/123123",
    },
  ],
]
`;

exports[`Remove grid suite Handles successful query Calling handler Invoked callback 1`] = `
Array [
  Array [
    null,
    Object {
      "grid": Object {
        "uuid": "123123",
      },
      "result": "mock",
    },
  ],
]
`;

exports[`Remove grid suite Handles unsuccesful response  Calling handler Returned the error to the callback 1`] = `
Array [
  Array [
    [StatusCodeError: 404 - ""],
  ],
]
`;
